hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/StreamToStreamJoinPhysicalRule.java

```diff
@@ -0,0 +1,397 @@
+     */
+    private WatermarkedFields watermarkedFields(
+            JoinLogicalRel join,
+            WatermarkedFields leftFields,
+            WatermarkedFields rightFields) {
+        final int offset = join.getLeft().getRowType().getFieldList().size();
+        Set<Integer> shiftedRightProps = rightFields.getFieldIndexes()
+                .stream()
+                .map(right -> right + offset)
+                .collect(Collectors.toSet());
+
+        return leftFields.union(new WatermarkedFields(shiftedRightProps));
+    }
+
+    /**
+     * Checks if all watermarked fields are temporal type.
+     */
+    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
+    private boolean watermarkedFieldsAreTemporalTypeCheck(RelNode input, WatermarkedFields fields) {
+        for (Integer idx : fields.getFieldIndexes()) {
+            RelDataType type = input.getRowType().getFieldList().get(idx).getType();
+            if (!HazelcastTypeUtils.isTemporalType(type)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+
+    @SuppressWarnings("CheckStyle")
+    private static final class BoundsExtractorVisitor extends RexVisitorImpl<Void> {
+        private final WatermarkedFields watermarkedFields;
+        private final String defaultErrorMessage = "Stream-to-stream JOIN condition must contain time boundness predicate";
+        private String errorMessage = null;
+
+        public Tuple3<Integer, Integer, Long> leftBound = null;
+        public Tuple3<Integer, Integer, Long> rightBound = null;
+
+        public boolean isValid = true;
+
+        private BoundsExtractorVisitor(WatermarkedFields watermarkedFields) {
+            super(true);
+            this.watermarkedFields = watermarkedFields;
+        }
+
+        public String errorMessage() {
+            return errorMessage == null ? defaultErrorMessage : errorMessage;
+        }
+
+        // Note: we're expecting equality or `a BETWEEN b and c` call here. Anything else will be rejected.
+        @Override
+        public Void visitCall(RexCall call) {
+            switch (call.getKind()) {
+                case AND:
+                    return super.visitCall(call);
+                case EQUALS:
+                    List<RexNode> operands = call.getOperands();
+                    if (operands.get(0) instanceof RexInputRef && operands.get(1) instanceof RexInputRef) {
+                        RexInputRef leftOp = (RexInputRef) operands.get(0);
+                        RexInputRef rightOp = (RexInputRef) operands.get(1);
+                        leftBound = tuple3(leftOp.getIndex(), rightOp.getIndex(), 0L);
+                        rightBound = tuple3(rightOp.getIndex(), leftOp.getIndex(), 0L);
+                    } else {
+                        isValid = false;
+                        errorMessage = "Time boundness or time equality condition are supported for stream-to-stream JOIN";
```

Comment:
Shouldn't it be "not supported"?

Reply:
I'd say, "**Only** time boundness or time equality ... "

Reply:
Oh, ok